### PR TITLE
feat: type inference for generators

### PR DESCRIPTION
**Pull request by lowr**
This PR implements basic type inference for generator and yield expressions.

Things not included in this PR:
- Generator upvars and generator witnesses are not implemented. They are only used to determine auto trait impls, so basic type inference should be fine without them, but method resolutions with auto trait bounds may not be resolved correctly.

Open questions:
- I haven't (yet) implemented `HirDisplay` for `TyKind::Generator`, so generator types are just shown as "{{generator}}" (in tests, inlay hints, hovers, etc.), which is not really nice. How should we show them?
- I added moderate amount of stuffs to minicore. I especially didn't want to add `impl<T> Deref for &T` and `impl<T> Deref for &mut T` exclusively for tests for generators; should I move them into the test fixtures or can they be placed in minicore?

cc #4309

### CODE REVIEW

**Comment by lowr** [comment]
I deliberately did not link #4309 for automatic close because there are still some features missing as mentioned in the PR description. However I think this PR enables type inference in most of the real-world use cases and we might want to close #4309 and create a new issue for the missing features (only if this PR is merged, of course).

**Comment by Veykril** [comment]
> I added moderate amount of stuffs to minicore. I especially didn't want to add impl<T> Deref for &T and impl<T> Deref for &mut T exclusively for tests for generators; should I move them into the test fixtures or can they be placed in minicore?

I think those are some key implementations so adding them to minicore seems fine to me

> I haven't (yet) implemented HirDisplay for TyKind::Generator, so generator types are just shown as "{{generator}}" (in tests, inlay hints, hovers, etc.), which is not really nice. How should we show them?

Probably just like closures but maybe with `gen` prefixed or something along those lines?

File: crates/hir-ty/src/infer.rs [hunk 1]
```diff
@@ -412,6 +412,7 @@ pub(crate) struct InferenceContext<'a> {
     /// closures, but currently this is the only field that will change there,
     /// so it doesn't make sense.
     return_ty: Ty,
+    resume_yield_tys: Option<(Ty, Ty)>,
```
Comment (Veykril) [inline crates/hir-ty/src/infer.rs:415]: Comment might be nice about which one the yield and return type is

**Comment by lowr** [comment]
> > I haven't (yet) implemented HirDisplay for TyKind::Generator, so generator types are just shown as "{{generator}}" (in tests, inlay hints, hovers, etc.), which is not really nice. How should we show them?
> 
> Probably just like closures but maybe with `gen` prefixed or something along those lines?

I'd love to reuse the closure syntax, but I'm not sure how we can encode yield type of a generator. Another option I thought of was `impl Generator<ResumeType, Yield = YieldType, Return = ReturnType>`, it's clear but too long to show imo (maybe it's suitable for `DisplayTarget::Test` though).

File: crates/hir-ty/src/infer.rs (hunk 1 above)
Comment (lowr) [reply crates/hir-ty/src/infer.rs:416]: Added doc comment

**Comment by lowr** [comment]
Sorry for the delay! I decided to go for the initial implementation like this: `|ResumeType| yields YieldType -> ReturnType`.

This is now ready for (another?) full review.

**Pull request merged**